### PR TITLE
Support listing an individual bean

### PR DESCRIPTION
**Pull request by skarsaune**
Signed-off-by: marska <martin.skarsaune@kantega.no>

List requests previously listed all beans in context even if one single bean is addressed in the path e.g. spring@domain/name=beanName

### CODE REVIEW

File: support/spring/src/main/java/org/jolokia/support/spring/backend/SpringListHandler.java [hunk 1]
```diff
@@ -44,10 +45,34 @@ public Object handleRequest(JolokiaListRequest pJmxReq, Object pPreviousResult)
...
+    /**
+     * Try to match up the path segment of the request to a spring bean
+     * @return Bean definition corresponding to request, null if none
+     */
+    private BeanDefinition beanFromRequest(JolokiaListRequest pJmxReq, String providerAndDomain) {
+        List<String> pathParts = pJmxReq.getPathParts();
+        if(pathParts != null && pathParts.size() == 2 && providerAndDomain.equals(
+                pathParts.get(0))) {
+           final String beanAndName = pathParts.get(1);
+           if(beanAndName.startsWith("name=")) {
```
Comment (rhuss) [inline support/spring/src/main/java/org/jolokia/support/spring/backend/SpringListHandler.java]: Maybe `beanAndName.toLowerCase().startsWith("name=")` ?

File: support/spring/src/main/java/org/jolokia/support/spring/backend/SpringListHandler.java [hunk 2]
```diff
@@ -44,10 +45,34 @@ public Object handleRequest(JolokiaListRequest pJmxReq, Object pPreviousResult)
...
+               final String beanName=beanAndName.substring(5);
```
Comment (rhuss) [inline support/spring/src/main/java/org/jolokia/support/spring/backend/SpringListHandler.java]: ... substring("name=".length()) and maybe move the string to a constant ?

File: support/spring/src/main/java/org/jolokia/support/spring/backend/SpringListHandler.java (hunk 1 above)
Comment (skarsaune) [inline support/spring/src/main/java/org/jolokia/support/spring/backend/SpringListHandler.java]: Done

File: support/spring/src/main/java/org/jolokia/support/spring/backend/SpringListHandler.java (hunk 2 above)
Comment (skarsaune) [inline support/spring/src/main/java/org/jolokia/support/spring/backend/SpringListHandler.java]: Done

**Comment by skarsaune** [comment]
@rhuss : current implementation was based on the hawtio JXM attribute roundtrip, where the JMX tree is initially built by getting all beans, and then each individual bean is queried with the complete path including name= on viewing attributes to decide data type, readability etc.

Could you provide an example of other urls patterns that the handler should support?

**Comment by rhuss** [comment]
Actual the idea is that everything behind the `/list/` is considered to be a path applied on the full return value, much like JSONPath. So these path elements should be actually really values of the tree return by a full `/list`. See also for the JMX example.

The 'name=' thing in Hawtio is a bit misleading as its part of the JMX MBean objectname, not something specific to Jolokia.

Important to support would be the queryparameter `maxDepth` (which cuts of the tree to return).

From my understanding hawt.io does the following:

* Call `/list?maxDepth=2` to get a tree with the JMX domain and MBean name.
* Then for an individual MBean it uses `/list/<domain>/<key-value-list>` from the tree obtained above. the `<key-value-list>` has the typical ObjectName form of key-value pairs, separated by commas. In the spring case this would be simply the spring bean name. 

hope this all makes sense ;-)

**Comment by skarsaune** [comment]
I think I understand, but it seems that in the top level list request jolokia already uses name= prefix for the spring beans it returns:

`for (String beanName : appCtx.getBeanDefinitionNames()) {		              
              BeanDefinition bd = bdFactory.getMergedBeanDefinition(beanName);
               ret.put("name=" + beanName, getSpringBeanInfo(bd));		 
          }`

So if one should address a path in the JSON tree, then we must use the same name?

Since jolokia supports multiple backends, it would look like it is a good idea to use a naming convention that is valid across backends?  it seems that the Jolokia JMX requests also have some requirements for naming conventions. For instance, if I test with Spring beans that have names that do not make valid objectName (e.g. with a colon) I get this:

java.lang.IllegalArgumentException: Invalid path within the MBean part given. (Path: spring@default/name=constretto:constrettoConfiguration)
	at org.jolokia.service.jmx.handler.ListHandler.doHandleAllServerRequest(ListHandler.java:92)
	at org.jolokia.service.jmx.handler.ListHandler.doHandleAllServerRequest(ListHandler.java:41)
	at org.jolokia.service.jmx.handler.AbstractCommandHandler.handleAllServerRequest(AbstractCommandHandler.java:69)
	at org.jolokia.service.jmx.LocalRequestHandler.handleRequest(LocalRequestHandler.java:77)
	at org.jolokia.server.core.backend.RequestDispatcherImpl.dispatch(RequestDispatcherImpl.java:46)
	at org.jolokia.server.core.backend.BackendManager.callRequestDispatcher(BackendManager.java:166)
	at org.jolokia.server.core.backend.BackendManager.handleRequest(BackendManager.java:94)
	at org.jolokia.server.core.http.HttpRequestHandler.executeRequest(HttpRequestHandler.java:196)
	at org.jolokia.server.core.http.HttpRequestHandler.handlePostRequest(HttpRequestHandler.java:136)
	at org.jolokia.jvmagent.handler.JolokiaHttpHandler.executePostRequest(JolokiaHttpHandler.java:224)
	at org.jolokia.jvmagent.handler.JolokiaHttpHandler.doHandle(JolokiaHttpHandler.java:162)
	at org.jolokia.jvmagent.handler.JolokiaHttpHandler.handle(JolokiaHttpHandler.java:101)
	at com.sun.net.httpserver.Filter$Chain.doFilter(Filter.java:79)
	at sun.net.httpserver.AuthFilter.doFilter(AuthFilter.java:83)
	at com.sun.net.httpserver.Filter$Chain.doFilter(Filter.java:82)
	at sun.net.httpserver.ServerImpl$Exchange$LinkHandler.handle(ServerImpl.java:675)
	at com.sun.net.httpserver.Filter$Chain.doFilter(Filter.java:79)
	at sun.net.httpserver.ServerImpl$Exchange.run(ServerImpl.java:647)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: javax.management.MalformedObjectNameException: Invalid character ':' in value part of property
	at javax.management.ObjectName.construct(ObjectName.java:618)
	at javax.management.ObjectName.<init>(ObjectName.java:1382)
	at org.jolokia.service.jmx.handler.ListHandler.objectNameFromPath(ListHandler.java:134)
	at org.jolokia.service.jmx.handler.ListHandler.doHandleAllServerRequest(ListHandler.java:79)
	... 20 more
08:14:53,649 INFO  [Thread-14] jolokia - Response: {"request":{"path":"spring@default\/name=constretto:constrettoConfiguration","type":"list"},"stacktrace":"java.lang.IllegalArgumentException: Invalid path within the MBean part given. (Path: spring@default\/name=constretto:constrettoConfiguration)\n\tat org.jolokia.service.jmx.handler.ListHandler.doHandleAllServerRequest(ListHandler.java:92)\n\tat org.jolokia.service.jmx.handler.ListHandler.doHandleAllServerRequest(ListHandler.java:41)\n\tat org.jolokia.service.jmx.handler.AbstractCommandHandler.handleAllServerRequest(AbstractCommandHandler.java:69)\n\tat org.jolokia.service.jmx.LocalRequestHandler.handleRequest(LocalRequestHandler.java:77)\n\tat org.jolokia.server.core.backend.RequestDispatcherImpl.dispatch(RequestDispatcherImpl.java:46)\n\tat org.jolokia.server.core.backend.BackendManager.callRequestDispatcher(BackendManager.java:166)\n\tat org.jolokia.server.core.backend.BackendManager.handleRequest(BackendManager.java:94)\n\tat org.jolokia.server.core.http.HttpRequestHandler.executeRequest(HttpRequestHandler.java:196)\n\tat org.jolokia.server.core.http.HttpRequestHandler.handlePostRequest(HttpRequestHandler.java:136)\n\tat org.jolokia.jvmagent.handler.JolokiaHttpHandler.executePostRequest(JolokiaHttpHandler.java:224)\n\tat org.jolokia.jvmagent.handler.JolokiaHttpHandler.doHandle(JolokiaHttpHandler.java:162)\n\tat org.jolokia.jvmagent.handler.JolokiaHttpHandler.handle(JolokiaHttpHandler.java:101)\n\tat com.sun.net.httpserver.Filter$Chain.doFilter(Filter.java:79)\n\tat sun.net.httpserver.AuthFilter.doFilter(AuthFilter.java:83)\n\tat com.sun.net.httpserver.Filter$Chain.doFilter(Filter.java:82)\n\tat sun.net.httpserver.ServerImpl$Exchange$LinkHandler.handle(ServerImpl.java:675)\n\tat com.sun.net.httpserver.Filter$Chain.doFilter(Filter.java:79)\n\tat sun.net.httpserver.ServerImpl$Exchange.run(ServerImpl.java:647)\n\tat java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)\n\tat java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)\n\tat java.lang.Thread.run(Thread.java:745)\nCaused by: javax.management.MalformedObjectNameException: Invalid character ':' in value part of property\n\tat javax.management.ObjectName.construct(ObjectName.java:618)\n\tat javax.management.ObjectName.<init>(ObjectName.java:1382)\n\tat org.jolokia.service.jmx.handler.ListHandler.objectNameFromPath(ListHandler.java:134)\n\tat org.jolokia.service.jmx.handler.ListHandler.doHandleAllServerRequest(ListHandler.java:79)\n\t... 20 more\n","error_type":"java.lang.IllegalArgumentException","error":"java.lang.IllegalArgumentException : Invalid path within the MBean part given. (Path: spring@default\/name=constretto:constrettoConfiguration)","status":400}

**Comment by rhuss** [comment]
yes, you are right (and I was wrong ;-)

The names should be all valid object names. You are right, as we construct the name on the fly for the spring handler witht he `name=` prefix, it should be also used in the path. So you original solution is correct. 

I wonder whether we should sanitize Spring bean names to be valid ObjectNames ? I guess this object name restriction is tightly coupled into the core of Jolokia (not sure).

**Comment by skarsaune** [comment]
One can note that down as a separate issue. The thoughtwork I guess is how to do mapping both ways, if some characters are valid in a spring bean name but not in an ObjectName.

**Comment by rhuss** [comment]
Yes, lets do it in a separate task.

**Comment by rhuss** [comment]
Just opened #348 for the sanitization task. Let's merge now ...

**Pull request merged**